### ISSUE TITLE
ui-grid 3.0.0-rc16: Unable to explicitly mark a row as "errored" with RowEdit

### ISSUE THREAD

**Issue by pmariduena** [angular-ui/ng-grid#2537]
I'm trying to press a "DELETE ROW" button in the UI, that calls in turn some code in my controller, which in turns makes a $resource REST call to the appropriate Server-side endpoint to actually delete the data corresponding to the row on the database.

My issue is that, if a HTTP 500 (or other error) is returned, and I'm using rowEditWaitInterval = -1, I am unable to mark the row as "Errored", or bringht red. This is what i'm trying to do the in the code to no avail:

```
// invoked after selecting a row to delete in the UI
$scope.deleteRow = function(rowEntity) {
// myResourceService is defined in another file, and is $resource-based, like this:
// $resource('app/myRows/:id/:otherId', {id: '@id', otherId: '@otherId'}, {
//                     
//             {
...
//                      removeRow: {method: 'DELETE'}
//             });
//
//
// rowToDeleteIndex has the index of the row to delete.
...
                    $log.log("Row number to delete is: " + rowToDeleteIndex);
                    var deferred = myResourceService.removeRow({
                        id: rowEntity.id,
                        otherId: rowEntity.otherId
                    });

                    // explicitly create a deferred object to be able to reject.
                    var rowDeferred = $q.defer();

                    // register save promise with the row to delete.
                    $scope.gridApi.rowEdit.setSavePromise($scope.gridApi.grid, rowEntity, rowDeferred.promise);

                    deferred.$promise.then(function (response) {
                        var newLengthOfGrid = $scope.myOverridesData.data.posts.splice(rowToDeleteIndex, 1);
                        $log.log('Successfully deleted row, value: Response status: ' + angular.toJson(response));
                        rowDeferred.resolve(newLengthOfGrid);
                    }, function (response) {
                        $log.error('Failed to deleted row: ' + angular.toJson(response) + ". status code: " + response.status
                        + '. statusText: ' + response.statusText + ', message: ' + response.data.message);
                        //deferred.reject(response.data);
                        $scope.warningMessage = response.data.message;
                        $scope.operationFailedFlag = true;
                        rowDeferred.reject(response.data);
                    });
                    $log.log('finished calling deleteRow');
}
```

**Comment by pmariduena** [comment]
After all the above code, I do receive the expected error on the server-side, however, the row is not marked bright red as expected.

Thanks.

**Comment by PaulL1** [comment]
I think this would be easier to diagnose with a plunker.  Is that possible, perhaps one based on the rowEdit tutorial?

**Comment by pmariduena** [comment]
Here's a plunker:

http://plnkr.co/spDnNxpQqvDhQrhY7VzL

However, I'm using a bogus URL against which to test the delete (since I don't know of one that actually can remove data). In any case, you can either select a row (and hit Remove Row button), or you can edit a row and mark it is "male" and hit "Save Row" and either should invoke the intended action, which is to explicitly reject() a deferred, such that the registered row (setSavePromise) is marked as "bright red" with error. But it's not happening.

**Comment by pmariduena** [comment]
The main test here is with the delete button.

**Comment by pmariduena** [comment]
For some reason, the $resource call is going to an OPTIONS verb instead of the intended DELETE action, but either way, it's still invoking the failure callback, which is the path we want to test on a Remove Row operation.

**Comment by PaulL1** [comment]
OK, the problem here is that your sequence isn't initially triggered by a save - so rowEdit isn't really set up to work with deletes.

The lifecycle is:
- request save on a row
- the `rowEdit.on.saveRow` event is raised, inside which you need to set a promise
- the save row code then waits on that promise, setting the row to error if the promise fails

The problem is that you're setting the promise, but you've never called saveRow, so nothing is waiting on that promise.

I suspect you could get a better result by changing your removeRow function to:
- set a flag on the row - perhaps `row.entity.delete = true`
- call saveRow, which in turn will call the `saveRow` event
- inside your event handler, check for the delete flag, if it's set then call remove instead of save
- set the promise based on that

Something like this: http://plnkr.co/edit/mv2mBMliKk2jkCmQ3ec9?p=preview

I also think you're kind of using resource wrong, so I changed to something I'd see as more standard - where all the data entities are resources so you don't have to mess around with the indexOf logic.

**Comment by pmariduena** [comment]
I'll try this out soon in my application.

**Comment by pmariduena** [comment]
I tried out your plunkr, but it doesn't work....here's the link, and the error I get when I try to delete a row (doesn't proceed to what we were trying to demo (the marking of a grid as dirty on error):

http://plnkr.co/edit/mv2mBMliKk2jkCmQ3ec9

Error:
```
selected row: 5
angular.js:9778 selected row name is: Madge Wilkerson
angular.js:9778 TypeError: undefined is not a function
    at Object.service.setRowsDirty (ui-grid-unstable.js:17741)
    at Grid.service.initializeGrid.publicApi.methods.rowEdit.setRowsDirty (ui-grid-unstable.js:17360)
    at Object.module.service.s.createBoundedWrapper [as setRowsDirty] (ui-grid-unstable.js:8743)
    at Scope.$scope.removeRow (script.js:78)
    at Parser.functionCall (angular.js:10567)
    at angular.js:18627
    at Scope.$get.Scope.$eval (angular.js:12412)
    at Scope.$get.Scope.$apply (angular.js:12510)
    at HTMLButtonElement.<anonymous> (angular.js:18626)
    at angular.js:2780angular.js:9778 (anonymous function)angular.js:7216 $getangular.js:12512 $get.Scope.$applyangular.js:18626 (anonymous function)angular.js:2780 (anonymous function)angular.js:330 forEachangular.js:2779 eventHandler
```

**Comment by pmariduena** [comment]
I'm going to cancel this...things seem to be working in the delete space for me now.

**Comment by pmariduena** [comment]
See #2729 looks like this was my issue here as well.
